[MLIR][LLVM] Add Continuous Loop Peeling transform to SCF

muneebkhan85 opened the pull request:
This patch adds continuous loop peeling to scf loop transforms in the MLIR backend. This transforms the target loop into a chain of loops, with step sizes that are powers of two and decrease exponentially across subsequent loops. 

The transform is similar to loop.peel in the effect that it creates a loop with a step (that is power of 2) to divide the range evenly, with the difference that the remaining iterations are spread across similar loops with exponentially decreasing step sizes, with the last loop with step size of 2^0 = 1.

Originally authored by Litu Zhou litu.zhou@huawei.com.

MaheshRavishankar reviewed (dismissed):
Could you also just create a separate C++ utility function for the core transformation (outside of the transform dialect). Thanks!

dcaballe commented:
Fantastic! I was thinking about this transformation recently.

> Could you also just create a separate C++ utility function for the core transformation (outside of the transform dialect). Thanks!

You can probably add the transformation itself to `mlir/lib/Dialect/SCF/Transforms/`. We have `LoopSpecialization` there which is also applying a peeling strategy

ThomasRaoux commented on mlir/test/Dialect/SCF/loop-continuous-peel.mlir:
```diff
@@ -0,0 +1,98 @@
+#map1 = affine_map<(d0) -> (d0)>
+module {
+  func.func @foo(%arg0: f32, %arg1: tensor<?xf32>) -> tensor<?xf32> {
+    %0 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel"]} ins(%arg0 : f32) outs(%arg1 : tensor<?xf32>) {
+    ^bb0(%in: f32, %out: f32):
+      %3 = arith.mulf %in, %out : f32
+      linalg.yield %3 : f32
+    } -> tensor<?xf32>
+    return %0 : tensor<?xf32>
```
Can you make the test more elementary? Ideally it should start directly from a loop and not depend on linalg. This will avoid dependency on tiling/linalg that are completely independent of this.

muneebkhan85 replied:
I have changed the test now, it works using scf.for and tests the transform in isolation.

muneebkhan85 commented:
> Fantastic! I was thinking about this transformation recently.
> 
> > Could you also just create a separate C++ utility function for the core transformation (outside of the transform dialect). Thanks!
> 
> You can probably add the transformation itself to `mlir/lib/Dialect/SCF/Transforms/`. We have `LoopSpecialization` there which is also applying a peeling strategy

I have moved the transform to mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp where I have also added the logic to expose this as an independent pass invoked with -scf-for-loop-continuous-peeling. I have also modified the test to use this method instead.

muneebkhan85 commented:
> Could you also just create a separate C++ utility function for the core transformation (outside of the transform dialect). Thanks!

Are you suggesting that I expose this as an independent pass? I have now done this and can be invoked using -scf-for-loop-continuous-peeling. The included test uses this method now.

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,168 @@ static void specializeForLoopForUnrolling(ForOp op) {
+          b, clonedOp, currentLoop.getInductionVar(), initialUb, initialStep,
+          /*insideLoop=*/true);
+      if (result.succeeded())
+        b.replaceOp(affineOp, currentLoop.getStep());
```
Is `replaceOp` and `eraseOp` needed? I think you can just run `rewritePeeledMinMaxOp` on the original `AffineMax/MinOp`. Is it possible to reuse `rewriteAffineOpAfterPeeling` here?

muneebkhan85 replied:
The step needs to be corrected once the rewritePeeledMinMaxOp has rewritten the affineOp. If not corrected, the old step from the original loop is propagated which is not correct in case of continuously peeling. This is what happens in the case of using rewriteAffineOpAfterPeeling. The generated code for the test case looks as shown below (see commented lines to spot the problem). Also, affineOp if used in scf::rewritePeeledMinMaxOp is not accessible after the function returns successfully, as the operation is rewritten by canonicalizeMinMaxOp. That is the reason for applying the rewriting on the clonedOp and then correcting the step using affineOp. Couldn't think if a better way of doing this.

```
#map = affine_map<()[s0, s1, s2] -> (s1 - s1 mod s2)>
#map1 = affine_map<(d0, d1)[s0] -> (8, d0 - d1)>  // step size 8 propogated as a constant. Incorrect assumption
                                                                                  // for continuous loop peeling
module {
  func.func @foo(%arg0: index) -> index {
    %c8 = arith.constant 8 : index
    %c4 = arith.constant 4 : index
    %c2 = arith.constant 2 : index
    %c0 = arith.constant 0 : index
    %0 = affine.apply #map()[%c0, %arg0, %c8]
    %1 = scf.for %arg1 = %c0 to %0 step %c8 iter_args(%arg2 = %c0) -> (index) {
      %10 = affine.min #map1(%arg0, %arg1)[%c8]
      %11 = index.add %10, %arg2
      scf.yield %11 : index
    }
    %2 = affine.apply #map()[%0, %arg0, %c4]
    %3 = arith.cmpi slt, %0, %2 : index
    %4 = scf.if %3 -> (index) {
      %10 = affine.min #map1(%arg0, %0)[%c8]  // step size not corrected!
      %11 = index.add %10, %1
      scf.yield %11 : index
    } else {
      scf.yield %1 : index
    }
    %5 = affine.apply #map()[%2, %arg0, %c2]
    %6 = arith.cmpi slt, %2, %5 : index
    %7 = scf.if %6 -> (index) {
      %10 = affine.min #map1(%arg0, %2)[%c8]  // step size not corrected!
      %11 = index.add %10, %4
      scf.yield %11 : index
    } else {
      scf.yield %4 : index
    }
    %8 = arith.cmpi slt, %5, %arg0 : index
    %9 = scf.if %8 -> (index) {
      %10 = affine.min #map1(%arg0, %5)[%c8]  // step size not corrected!
      %11 = index.add %10, %7
      scf.yield %11 : index
    } else {
      scf.yield %7 : index
    }
    return %9 : index
  }
}
```

muneebkhan85 commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,168 @@ static void specializeForLoopForUnrolling(ForOp op) {
   op.erase();
 }
 
+static LogicalResult splitLoopHelper(RewriterBase &b, scf::ForOp &forOp,
```
Yes, the function can be called from peelForLoop. The test mlir/test/Dialect/SCF/for-loop-peeling.mlir passes in either cases when spliLoopHelper is called in the begging or in the end as shown here

muneebkhan85 replied:
Yes, the function can be called from peelForLoop. The test mlir/test/Dialect/SCF/for-loop-peeling.mlir passes when spliLoopHelper is called after the new splitBound has been computed as shown here

`static LogicalResult peelForLoop(RewriterBase &b, ForOp forOp,
                                 ForOp &partialIteration, Value &splitBound) {

  auto loc = forOp.getLoc();
  AffineExpr sym0, sym1, sym2;
  bindSymbols(b.getContext(), sym0, sym1, sym2);
  // New upper bound: %ub - (%ub - %lb) mod %step
  auto modMap = AffineMap::get(0, 3, {sym1 - ((sym1 - sym0) % sym2)});
  b.setInsertionPoint(forOp);
  splitBound = b.createOrFold<AffineApplyOp>(loc, modMap,
                                             ValueRange{forOp.getLowerBound(),
                                                        forOp.getUpperBound(),
                                                        forOp.getStep()});

  if(failed(splitLoopHelper(b, forOp, partialIteration, splitBound)))
    return failure();

  return success();
}`

But, should this change be a part of this patch? I think this is a separate change, as the current patch should only target the new addition of Continuous Loop Peeling?

muneebkhan85 commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,168 @@ static void specializeForLoopForUnrolling(ForOp op) {
   op.erase();
 }
 
+static LogicalResult splitLoopHelper(RewriterBase &b, scf::ForOp &forOp,
```
Done.

muneebkhan85 commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,168 @@ static void specializeForLoopForUnrolling(ForOp op) {
+  return success();
+}
+
+static scf::IfOp convertSingleIterFor(RewriterBase &b, scf::ForOp &forOp) {
```
Done.

muneebkhan85 commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,168 @@ static void specializeForLoopForUnrolling(ForOp op) {
+  // then branch
+  b.setInsertionPointToStart(ifOp.thenBlock());
+  for (Operation &op : forOp.getBody()->getOperations()) {
+    b.clone(op, mapping);
```
This is a good suggestion, but it seems like a more involved change. I can't address this atm but I could add a TODO comment about this.

muneebkhan85 replied:
Done. Replaced cloning of operations with using RewriterBase::inlineBlockBefore to move the operations from the scf::for (to be removed) to the newly created scf::if.

muneebkhan85 replied:
Done.

muneebkhan85 commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,168 @@ static void specializeForLoopForUnrolling(ForOp op) {
+  auto stepInt = getConstantIntValue(forOp.getStep());
+
+  // Step size must be a known positive constant greater than 1.
+  if (stepInt && stepInt <= static_cast<int64_t>(1))
```
Yes, thats correct, but stepInt should be checked against 1. As is the case for peelForLoop, this transform is also not to be applied when step is 1 or 0. Latest changes look as follows

```
  // Step size must be a known positive constant greater than 1.
  if (!stepInt || stepInt <= static_cast<int64_t>(1))
    return failure();
```

muneebkhan85 commented:
ping @matthias-springer

matthias-springer commented on mlir/include/mlir/Dialect/SCF/TransformOps/SCFTransformOps.td:
```diff
@@ -147,6 +147,43 @@ def LoopPeelOp : Op<Transform_Dialect, "loop.peel",
+     TransformOpInterface, TransformEachOpTrait]> {
+  let description = [{
+    Transforms the loop into a chain of loops, with step sizes that are
+    powers of two and decrease exponetially across subsequent loops.
```
typo

matthias-springer commented on mlir/include/mlir/Dialect/SCF/Transforms/Transforms.h:
```diff
@@ -81,6 +81,47 @@ void naivelyFuseParallelOps(Region &region);
+///   (loop body)
+/// }
+/// %newUb2 = %ub - (%ub mod %c4)
+/// scf.for %iv2 = %newUb to %newUb2 {
```
Step size is missing.

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,165 @@ static void specializeForLoopForUnrolling(ForOp op) {
+  IRMapping mapping;
+  mapping.map(forOp.getInductionVar(), forOp.getLowerBound());
+  for (auto [arg, operand] :
+       llvm::zip(forOp.getRegionIterArgs(), forOp.getInitsMutable())) {
```
nit: `zip_equal`

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,165 @@ static void specializeForLoopForUnrolling(ForOp op) {
+    b.setInsertionPoint(currentLoop);
+    auto constStepOp =
+        b.create<arith::ConstantIndexOp>(currentLoop.getLoc(), loopStep);
+    currentLoop.getStepMutable().assign(constStepOp);
```
This should be wrapped in `updateRootInPlace` because existing operations may be modified.

matthias-springer reviewed (commented):
Sorry, I dropped the ball on this review. Here a few more small comments, I'm going to do another more thorough review.

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,167 @@ static void specializeForLoopForUnrolling(ForOp op) {
+  currentLoop = forOp;
+  AffineExpr sym0, sym1, sym2;
+  bindSymbols(b.getContext(), sym0, sym1, sym2);
+  AffineMap defaultSplitMap =
```
Add comment to explain why you have `defaultSplitMap` and `powerSplitMap`.

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,167 @@ static void specializeForLoopForUnrolling(ForOp op) {
+/// Create a new for loop for the remaining iterations (partiaIteration)
+/// after a for loop has been peeled. This is followed by correcting the
+/// loop bounds for both loops given the index (splitBound) where the
+/// iteration space is to be split up.
```
Explain the meaning of the return value.

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,167 @@ static void specializeForLoopForUnrolling(ForOp op) {
+      break;
+    currentLoop = partialIteration;
+    uint64_t maxPower = llvm::bit_floor(loopStep);
+    loopStep = maxPower == loopStep ? maxPower >> 1 : maxPower;
```
nit: try to avoid bit-shifting when the same logic can be expressed with multiplication/division. (so that the code is easier to read)

matthias-springer commented on mlir/test/Dialect/SCF/for-loop-continuous-peeling.mlir:
```diff
@@ -0,0 +1,46 @@
+func.func @foo(%ub: index) -> index {
+  %c0 = arith.constant 0 : index
+  %step = arith.constant 8 : index
+  %0 = scf.for %iv = %c0 to %ub step %step iter_args(%arg = %c0) -> (index) {
```
Add another test case so that both `usePowerSplit = false` and `usePowerSplit = true` is exercised.

matthias-springer commented on mlir/test/Dialect/SCF/for-loop-continuous-peeling.mlir:
```diff
@@ -0,0 +1,46 @@
+// CHECK: %[[I1:.*]] = scf.for %{{.*}} = %[[LB]] to %[[I0]] step %[[STEP8]] iter_args(%[[ALB:.*]] = %[[LB]]) -> (index) {
+// CHECK: %[[SUM:.*]] = index.add %[[ALB]], %[[STEP8]]
+// CHECK: scf.yield %[[SUM]] : index
+// CHECK: %[[I2:.*]] = affine.apply #[[MAP]]()[%[[I0]], %[[UB]], %[[STEP4]]]
```
Please indent for better readability.

matthias-springer commented on mlir/lib/Dialect/SCF/Utils/AffineCanonicalizationUtils.cpp:
```diff
@@ -189,16 +188,17 @@ LogicalResult scf::canonicalizeMinMaxOpInLoop(RewriterBase &rewriter,
 /// * Inside the partial iteration: min(step, ub - iv) == ub - iv
 ///
-/// Returns `success` if the given operation was replaced by a new operation;
+/// Returns the new Affine op if the operation was replaced by a new operation;
```
`new AffineApplyOp`

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,167 @@ static void specializeForLoopForUnrolling(ForOp op) {
+            b, affineOp, currentLoop.getInductionVar(), initialUb, initialStep,
+            /*insideLoop=*/true);
+        // correct the step of the newly created affine op
+        if (!failed(result))
```
`if (succeeded(result))`

matthias-springer commented on mlir/lib/Dialect/SCF/Transforms/LoopSpecialization.cpp:
```diff
@@ -105,6 +106,167 @@ static void specializeForLoopForUnrolling(ForOp op) {
+            /*insideLoop=*/true);
+        // correct the step of the newly created affine op
+        if (!failed(result))
+          b.replaceOp(result.value(), currentLoop.getStep());
```
I don't understand this part. There is no guarantee that `affineOp` looks like `min(ub - iv, 8)`. It could be an arbitrary affine.min/affine.max op.
